clang/lib/CIR/CodeGen/CIRGenItaniumCXXABI.cpp

```diff
@@ -2288,6 +2288,23 @@ static mlir::cir::FuncOp getItaniumDynamicCastFn(CIRGenFunction &CGF) {
   return CGF.CGM.createRuntimeFunction(FTy, "__dynamic_cast");
 }
 
+mlir::cir::DynamicCastToVoidInfoAttr
+CIRGenItaniumCXXABI::buildDynamicCastToVoidInfo(CIRGenFunction &CGF,
+                                                QualType SrcRecordTy) {
+  mlir::Type vtableElemTy;
```

Comment:
Adding an attribute seems not ideal, it just papers over the fact that we need more information available in LoweringPrepare. See specific comments below.

Reply:
Well, the problem here is dyn-cast-to-void requires completely different set of ABI-specific information than a normal downcast or sidecast. So if you take the attribute approach (i.e. you throw all interesting ABI-specific information into an attribute and attach that attribute to the operation) you can't avoid having two different kinds of attributes.

AST query during lowering prepare seems more promising in this situation. Both `#cir.dyn_cast_info` and `#cir.dyn_cast_to_void_info` are computed from the AST, thus we can delay their computation until lowering prepare. This further allows us to remove `#cir.dyn_cast_info` and `#cir.dyn_cast_to_void_info` completely and make `cir.dyn_cast` clearer in its assembly form.

Reply:
I think there might be value in keeping `#cir.dyn_cast_info` for now, it's already there and it encodes useful information making it easier to lower. More specifically, for keeping a direct reference to the symbols, it might be really handy for optimizations: e.g. it allows us to identify symbol usage/dependency, which is useful for deleting unused code (or globals).

`#cir.dyn_cast_to_void_info` is a different story IMO and could be implemented by querying the AST. I previously gave the suggestion of adding another kind, such as `DCK_PtrCastToVoid`, but couldn't we just look at the op result type to check whether it's a void pointer? I see this in `CIROps.td` definition:

```
let results = (outs StructPtr:$result);
```

Should this be a `StructOrVoidPtr` instead?

Reply:
> but couldn't we just look at the op result type to check whether it's a void pointer?

Yes of course this is possible. Instead of adding more stuff to the existing one, what about a new operation specifically for dyn-cast-to-void? Something like `cir.dyn_cast_to_void`.

Reply:
I'd prefer we encapsulate into just one operation, looks like this could be solved by adding a simpler version of your attribute: an `UnitAttr:$relative_layout`. This is all that lowering should look at, it should provide ABI answers just as accurate as during CIRGen.

Reply:
Updated.